sys/net/dsm/dsm.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2021 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_dsm
+ * @{
+ *
+ * @file
+ * @brief       DTLS Session Management module implementation
+ *
+ * @author      János Brodbeck <janos.brodbeck@ml-pa.com>
+ *
+ * @}
+ */
+
+#include "net/dsm.h"
+#include "mutex.h"
+#include "net/sock/util.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+typedef struct {
+    sock_dtls_t *sock;
+    sock_dtls_session_t session;
+    dsm_state_t state;
+    uint32_t last_used_sec;
+} dsm_session_t;
+
+static int _find_session(sock_dtls_t *sock, sock_dtls_session_t *to_find,
+                         dsm_session_t **session);
+
+static mutex_t _lock;
+static dsm_session_t _sessions[DTLS_PEER_MAX];
+static uint8_t _available_slots;
+
+void dsm_init(void)
+{
+    mutex_init(&_lock);
+    _available_slots = DTLS_PEER_MAX;
+}
+
+dsm_state_t dsm_store(sock_dtls_t *sock, sock_dtls_session_t *session,
+                      dsm_state_t new_state, bool restore)
+{
+    sock_udp_ep_t ep;
+    dsm_session_t *session_slot = NULL;
+    dsm_state_t prev_state = NO_SPACE;
+    mutex_lock(&_lock);
+
+    ssize_t res = _find_session(sock, session, &session_slot);
+    if (res != -1) {
```

Comment:
```suggestion
    if (res >= 0) {
```

adds a bit of future-proofing should `_find_session` ever get refactored to emit more error code

or even better

```suggestion
    if (res < 0) {
        DEBUG("dsm: no space for session to store\n");
        goto out;
    }
```

and we can reduce the level of nesting

Reply:
Done.